### PR TITLE
[NPU]DeepSeek-V3.2 support npu mlaprolog

### PR DESCRIPTION
**Pull request by lawtherWu**
## Motivation

we enabled the DSV3.2 to support npu mla_prolog_v3 to improve the performance

## Modifications

- Added support for the mla_prolog_v3 custom operator, which includes operations such as Q/KV LoRA, RoPE, Norm, and KVCache update, and contains significant Cube/Vector parallelism space.
- Extracted the calls of the MLAPO operator and the MLA-PROLOG operator into the function `npu_mla_preprocess()`.
- Adjusted the weight of the `fused_qkv_a_proj_with_mqa` layer to be compatible with the mla_prolog_v3 operator.

## Accuracy Tests

acc

## Benchmarking and Profiling

Before
p1
After
p2
Before(MTP)
p3
After(MTP)
p4
Profilling
p5

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [ ] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/srt/layers/quantization/unquant.py [hunk 1]
```diff
@@ -119,12 +124,15 @@ def create_weights(
...
             _amx_process_weight_after_loading(layer, ["weight"])
+        if _is_npu:
+            _mlaprolog_process_weight_after_loading(layer)
 
     def apply(
         self,
         layer: torch.nn.Module,
         x: torch.Tensor,
         bias: Optional[torch.Tensor] = None,
+        dynamic_scale: torch.Tensor = None,
```
Comment (Alcanderian) [inline python/sglang/srt/layers/quantization/unquant.py]: Please replace `dynamic_scale` with `kw_args`

File: python/sglang/srt/layers/linear.py [hunk 2]
```diff
@@ -253,10 +260,12 @@ def weight_loader(self, param: Parameter, loaded_weight: torch.Tensor):
         assert param.size() == loaded_weight.size()
         param.data.copy_(loaded_weight)
 
-    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+    def forward(
+        self, x: torch.Tensor, dynamic_scale: torch.Tensor = None
```
Comment (Alcanderian) [inline python/sglang/srt/layers/linear.py]: Do not add such parameter in the public interface.
Please replace `dynamic_scale` with `kw_args`

**Review by Alcanderian — DISMISSED** [review]

File: python/sglang/srt/layers/quantization/unquant.py (hunk 1 above)
Comment (lawtherWu) [reply python/sglang/srt/layers/quantization/unquant.py]: done

File: python/sglang/srt/layers/linear.py (hunk 2 above)
Comment (lawtherWu) [reply python/sglang/srt/layers/linear.py]: done

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py [hunk 3]
```diff
@@ -611,6 +611,9 @@ def forward_extend(
         k_rope: Optional[torch.Tensor] = None,
         topk_indices: Optional[torch.Tensor] = None,
     ):
+        if is_mla_preprocess_enabled():
+            # MLAPO and MLAPROLOG does saving kv_cache
```
Comment (ping1jing2) [inline python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: \# MLAPO and MLAPROLOG do save kv_cache

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py [hunk 4]
```diff
@@ -399,17 +422,58 @@ def forward_mlapo(self, positions, hidden_states, forward_batch, zero_allocator)
...
+        )
+
     def forward(self, positions, hidden_states, forward_batch, zero_allocator):
         assert self.quant_config and self.quant_config.get_name() == "modelslim"
         # route by `qkv_a_proj` quant type as MTP layers can be unquantized
         _is_w8a8 = (
             hasattr(self.qkv_a_proj.quant_method, "quant_config")
             and self.qkv_a_proj.quant_method.quant_config.get_name() == "modelslim"
         )
+        _is_mlaprolog = not hasattr(self.qkv_a_proj.quant_method, "quant_config")
```
Comment (ping1jing2) [inline python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: #14424 we will support different frameworks such as gptq/awq, so please ensure this condition can work properly in this scenario?

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py [hunk 5]
```diff
@@ -399,17 +422,58 @@ def forward_mlapo(self, positions, hidden_states, forward_batch, zero_allocator)
...
         if _is_w8a8:
             return self.forward_mlapo(
                 positions, hidden_states, forward_batch, zero_allocator
             )
+        elif _is_mlaprolog:
+            return self.forward_mlaprolog(positions, hidden_states, forward_batch)
         else:
             return self.forward_absorb_prepare_npu_rms_norm_cache(
```
Comment (ping1jing2) [inline python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py:476]: what's the difference between `elif` and `else`? and do we need `else` branch now?

File: python/sglang/srt/models/deepseek_v2.py [hunk 6]
```diff
@@ -1331,6 +1331,7 @@ def __init__(
                 bias=False,
                 quant_config=quant_config,
                 prefix=add_prefix("fused_qkv_a_proj_with_mqa", prefix),
+                q_lora_rank=self.q_lora_rank,
```
Comment (ping1jing2) [inline python/sglang/srt/models/deepseek_v2.py]: Generally speaking, it is not recommanded to modify the public API. or could you give some solid reasons to explain why you must make such modification

File: python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py (hunk 3 above)
Comment (lawtherWu) [reply python/sglang/srt/hardware_backend/npu/attention/ascend_backend.py]: done

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py (hunk 4 above)
Comment (lawtherWu) [reply python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: Thank you for your review comments. Line 465 indicates that `NPUFusedMLAPreprocess.forward()` is exclusively triggered for `modelslim` quantization frameworks. We revised the conditional judgment to better distinguish the entry logic for the MLAPROLOG and MLAPO branches.

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py (hunk 5 above)
Comment (lawtherWu) [reply python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py:476]: In the `elif` branch, the custom `mlaprolog` operator fuses all preparatory computations prior to the `MLA`. In the `else` branch, `npu_kv_rmsnorm_rope_cache` fuses the computations of RMSNorm, RoPE, and cache update. This `else` branch is typically triggered in the MTP layer, so we should retain it.

File: python/sglang/srt/models/deepseek_v2.py (hunk 6 above)
Comment (lawtherWu) [reply python/sglang/srt/models/deepseek_v2.py]: Thank you for your review comments. The mlaprolog operator’s input parameters comprise the `q_a_proj` and `kv_a_proj` weights, while the `q_lora_rank` is required to split the `fused_qkv_a_proj_with_mqa` weight into two distinct parts. For the above reason, I believe this modification is necessary.

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py [hunk 7]
```diff
@@ -92,6 +111,10 @@ def __init__(
         self.qk_nope_head_dim = qk_nope_head_dim  # 128
         self.qk_rope_head_dim = qk_rope_head_dim  # 64
         self.qk_head_dim = qk_nope_head_dim + qk_rope_head_dim
+        self.v_head_dim = v_head_dim
+        self.dequant_scale_w_uq_qr = self.q_b_proj.weight_scale.view(1, -1).to(
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: suggest: keep the naming convention for better readability

File: python/sglang/srt/layers/linear.py [hunk 8]
```diff
@@ -253,10 +260,13 @@ def weight_loader(self, param: Parameter, loaded_weight: torch.Tensor):
...
 
-    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+    def forward(
+        self, x: torch.Tensor, **kwargs
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
         bias = self.bias if not self.skip_bias_add else None
         assert self.quant_method is not None
-        output = self.quant_method.apply(self, x, bias)
+        dynamic_scale = kwargs.get("dynamic_scale", None)
+        output = self.quant_method.apply(self, x, bias, dynamic_scale=dynamic_scale)
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/linear.py]: this is not a good modification imo

it needs to change all `apply` methods for all quantizations, not only `unquant` and `linear_method_npu`, this will be a huge work to do

i would suggest packing `x` and `dynamic_scale` into a tuple temply to solve this  issue

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py [hunk 9]
```diff
@@ -11,6 +12,23 @@
     from sglang.srt.layers.quantization.base_config import QuantizationConfig
 
 
+def _mlaprolog_process_weight_after_loading(layer):
```
Comment (iforgetmyname) [inline python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: this postprocess looks great but introducing `q_lora_rank` as input is hardly acceptable, i think of two ways to avoid changing interface:
1. follows `forward_mlapo` that use `has_preprocess_weights` flag to trigger weight postprocess. we have warmups for graph/eager mode so this won't affect model performance. refer: https://github.com/sgl-project/sglang/pull/15381/changes#diff-3d7852a24e487554c793ccbb4c91145b2f723a12d4ee35b68bad9b6bd3b01c53L326.
2. change `deepseek_v2.py`, add `q_lora_rank` as a temporary member value for `fused_qkv_a_proj_with_mqa`.
```python
            self.fused_qkv_a_proj_with_mqa = ReplicatedLinear(
                self.hidden_size,
                self.q_lora_rank + self.kv_lora_rank + self.qk_rope_head_dim,
                bias=False,
                quant_config=quant_config,
                prefix=add_prefix("fused_qkv_a_proj_with_mqa", prefix),
            )
            # Changes start here:
            self.fused_qkv_a_proj_with_mqa.q_lora_rank = self.q_lora_rank
```

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py (hunk 7 above)
Comment (lawtherWu) [reply python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: done

File: python/sglang/srt/layers/linear.py (hunk 8 above)
Comment (lawtherWu) [reply python/sglang/srt/layers/linear.py]: Thank you for your review, and I have made modifications according to your suggestion.

File: python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py (hunk 9 above)
Comment (lawtherWu) [reply python/sglang/srt/hardware_backend/npu/attention/mla_preprocess.py]: I have modified it according to the first way.

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 10]
```diff
@@ -1010,7 +968,7 @@ def forward_npu(
         if self.alt_stream is not None:
             self.alt_stream.wait_stream(torch.npu.current_stream())
             with torch.npu.stream(self.alt_stream):
-                q = self.wq_b(q_lora)[
+                q = self.wq_b(q_lora, dynamic_scale=dynamic_scale)[
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py]: combine `q_lora` and `dynamic_scale` into a tuple and pass into `self.wq_b`

File: python/sglang/srt/layers/quantization/unquant.py [hunk 11]
```diff
@@ -125,6 +125,7 @@ def apply(
         layer: torch.nn.Module,
         x: torch.Tensor,
         bias: Optional[torch.Tensor] = None,
+        **kwargs,
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/quantization/unquant.py]: revert this change

File: python/sglang/srt/layers/linear.py [hunk 12]
```diff
@@ -253,9 +254,14 @@ def weight_loader(self, param: Parameter, loaded_weight: torch.Tensor):
         assert param.size() == loaded_weight.size()
         param.data.copy_(loaded_weight)
 
-    def forward(self, x: torch.Tensor) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+    def forward(
+        self, x: torch.Tensor, **kwargs
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/linear.py]: follows above change, `(q_lora, dynamic_scale)` will pass in as `x` here and there's on more accesses/changes towards `x` but only pass it into `self.quant_method.apply` where we can handle it in `linear_method_npu.py`

Comment (lawtherWu) [reply python/sglang/srt/layers/linear.py]: done

File: python/sglang/srt/layers/quantization/unquant.py (hunk 11 above)
Comment (lawtherWu) [reply python/sglang/srt/layers/quantization/unquant.py]: done

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py (hunk 10 above)
Comment (lawtherWu) [reply python/sglang/srt/layers/attention/nsa/nsa_indexer.py]: done

File: python/sglang/srt/layers/attention/nsa/nsa_indexer.py [hunk 13]
```diff
@@ -1126,7 +968,7 @@ def forward_npu(
         if self.alt_stream is not None:
             self.alt_stream.wait_stream(torch.npu.current_stream())
             with torch.npu.stream(self.alt_stream):
-                q = self.wq_b(q_lora)[
+                q = self.wq_b((q_lora, dynamic_scale))[
```
Comment (iforgetmyname) [inline python/sglang/srt/layers/attention/nsa/nsa_indexer.py]:
```suggestion
                q_lora = (q_lora, dynamic_scale) if dynamic_scale is not None else q_lora
                q = self.wq_b(q_lora)[
```
looks great but one more fix here
avoid pass in a tuple when unquanted

Comment (lawtherWu) [reply python/sglang/srt/layers/attention/nsa/nsa_indexer.py]: done

**Review by ping1jing2 — APPROVED** [review]

**Review by iforgetmyname — APPROVED** [review]
